Update asset inlining strategy

evoactivity opened the pull request:
Reverts #238 and uses importSync to properly require the file.

I'm having issues with the tests though.

Not sure what to do now so creating the PR so more eyes can look

Resolves #251

evoactivity commented:
When I update dependancies I end up breaking the Ember 3 tests.
Should we drop testing ember 3 since 4.4, 4.8 and 4.12 are the current LTS releases?

evoactivity commented:
Everything but embroider is now passing, I don't know if the embroider failures are an issue with qunit-dom or something else at the moment. Not sure how to further debug.

evoactivity commented:
This is the relevant error

```
WARNING in ../rewritten-packages/ember-svg-jar.5302b7ef/helpers/svg-jar.js 6:15-58
Module not found: Error: Can't resolve 'ember-svg-jar/inlined' in '/Users/liam/Work/GitHub/ember-svg-jar/packages/ember-svg-jar/node_modules/.embroider/rewritten-packages/ember-svg-jar.5302b7ef/helpers'
```
The svgs are inlined to es modules by broccoli from treeForAddon which isn't used by embroider as far as I'm aware.
So I don't understand how my previous PR ever made the embroider scenarios pass if treeForAddon is not being run.

patricklx commented on packages/ember-svg-jar/addon/helpers/svg-jar.js:
```diff
@@ -0,0 +1,19 @@
+
+function getInlineAsset(assetId) {
+  try {
+    return importSync(`ember-svg-jar/inlined/${assetId}`).default;
```
I don't think you need.default here

patricklx commented:
There is also this adapter doing some replacing:
https://github.com/embroider-build/embroider/blob/main/packages/compat/src/compat-adapters/ember-svg-jar.ts

patricklx commented:
treeForAddon should be run

patricklx commented:
Found this too: https://github.com/embroider-build/embroider/issues/650

Maybe you need to use relative path?

evoactivity commented:
I've pushed something that passed the test locally.

There error from embroider-safe I've seen locally but randomly on different tests.

Ideally I'd be able to 
```
if (macroCondition(isEmbroider()) {}
```
But I can't find anything on how I might acheive that.

evoactivity commented:
Ok, everything passes now but I'd really like some ideas on using a macroCondition instead of my nested try catches.

evoactivity commented:
btw, I tried leaving it as a require to let embroider patch it but that apparently isn't working.

patricklx commented:
Nice
Maybe if you Can make the inlined packager create an index.js file that re-exports all created files? Then you could directly import it with a relative path and it would work for classic build

evoactivity commented:
I might be misunderstanding but I'n not sure how that would help.

According to https://github.com/embroider-build/embroider/issues/650 the relative path imports just don't work in classic builds.

patricklx commented:
From what i undersood the relative paths do not work only if there is a dynamic part in it.

evoactivity commented:
I see, I'll give it a try.

patricklx commented on packages/ember-svg-jar/addon/helpers/svg-jar.js:
```diff
@@ -1,6 +1,8 @@
 import { helper } from '@ember/component/helper';
 import makeSVG from 'ember-svg-jar/utils/make-svg';
-import icons from '../inlined';
```
I think you needed `import * as icons from ...`

evoactivity replied:
I switched back to importSync but you're probably right

evoactivity commented:
So that works but I'm concerned about the implications of pushing all icons into 1 file and loading them all.

patricklx commented:
> So that works but I'm concerned about the implications of pushing all icons into 1 file and loading them all.

Well, ideally embroider apps would import the svg directly somewhere and pass it to the svg helper 

I don't think it's too big issue, another idea would be to export functions that call import sync and return the result

evoactivity commented:
> another idea would be to export functions that call import sync and return the result

I like this idea. Seems to work well locally. Pushed up a change, fingers crossed CI is happy too.

evoactivity commented:
I'm happy with this as the solution. Unless there are any objections I'll squash the commits.

@patricklx Thanks for you help on this :)

@jherdman If you're happy with this can you cut a release asap please

patricklx commented on packages/ember-svg-jar/lib/inline-packer.js:
```diff
@@ -65,6 +62,8 @@ class InlinePacker extends CachingWriter {
     saveToFile(
       path.join(outputPath, `index.js`),
       `import { importSync } from '@embroider/macros';\n` +
+        `export ${makeSafeFunctionName.toString()}` +
```
Another idea that came to mind:
Instead if named exports, you could export an object.
const obj = {
  [name]: func,
  ...
}
export default obj;

evoactivity replied:
Well that's another good idea!

evoactivity commented:
Copying from issue
---

So to sum up the changes made

- Update related dependencies
- Bump node version in workflows to 16 so fastboot recognises nullish coalescing operator
- Output an object of functions keyed by the asset id as an index.js inside the inlinded dir
- Import the generated index.js as an es module in the helper
- Look up the assetId function and run it to import the related svg
- Added test to ensure files with apostophes work

Other improvements to infra that could be made:

- Now that ember-release, canary and beta are on v5 might be worthwhile adding try scenarios for v4 LTS releases.
- The symbol injector seems flaky in tests, I added a waitFor with a 10 second timeout but it still times out randomly.

but these can be done in another issue/pr.

evoactivity commented:
Well, I promise the tests pass, I'm not going to add another commit just to re-run the tests now I've squashed.

jherdman Can you rerun the tests and release if you're happy?

jherdman commented:
@evoactivity applogies, I am no longer involved with this project.

evoactivity commented:
@jherdman Do you know who has write access and is able to release now?

enspandi commented:
Thanks for the fix @evoactivity , we are using this branch for our embroider app (on v3.1.2) :clap: :+1:

void-mAlex commented:
discussed during Embroider team meeting today and this looks like an ok solution that is not expected to break or cause more work going forward

evoactivity commented:
I've merged this in, I'll cut a new release a little later tonight.
